flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobMasterServiceLeadershipRunner.java

```diff
@@ -488,18 +523,21 @@ private boolean isRunning() {
         return state == State.RUNNING;
     }
 
-    private void runIfValidLeader(
-            UUID expectedLeaderId, Runnable action, String actionDescription) {
+    private <T> CompletableFuture<T> runIfValidLeader(
```

Comment:
nit: This method should be called `supplyIfValidLeader`. But with your current implementation, a dedicated `runIfValidLeader` would be still helpful. Hint: SupplierAsync-based and Runnable-based helper methods can be translated to each other in the following way:
```
private <T> CompletableFuture<T> supplyAsyncIfValidLeader(
            UUID expectedLeaderId,
            Supplier<CompletableFuture<T>> supplier,
            Supplier<CompletableFuture<T>> noLeaderFallback) {
        final CompletableFuture<T> resultFuture = new CompletableFuture<>();
        runIfValidLeader(
                expectedLeaderId,
                () -> FutureUtils.forward(supplier.get(), resultFuture),
                () -> FutureUtils.forward(noLeaderFallback.get(), resultFuture));

        return resultFuture;
    }

    private void runIfValidLeader(
            UUID expectedLeaderId, Runnable action, Runnable noLeaderFallback) {
        synchronized (lock) {
            if (isValidLeader(expectedLeaderId)) {
                action.run();
            } else {
                noLeaderFallback.run();
            }
        }
    }
```

Reply:
I've added these two methods.